### DISCUSSION TITLE
How to manage custom extensions with Docker?

### DISCUSSION THREAD

**Discussion by abdonrd** [directus/directus#17310] [Archive]
Following the [Directus Docker guide](https://docs.directus.io/self-hosted/docker-guide.html), we have a project with the `docker-compose.yml`.
Internally it uses a `Dockerfile` (`build: .`), which is the `Dockerfile` we will use in production.

I made [this repository](https://github.com/abdonrd/directus-docker-extensions-demo) as a little demo with just two commits:

1. [`ebb6c8c`](https://github.com/abdonrd/directus-docker-extensions-demo/commit/ebb6c8c207ce9e3dda69a3710a99b7bd99fcaeab) Setup simple Directus project with docker compose
2. [`11d454d`](https://github.com/abdonrd/directus-docker-extensions-demo/commit/11d454df73fee12b024d53b42ca6d19c3a592381) Add a custom extension

We use [npm workspaces](https://docs.npmjs.com/cli/v9/using-npm/workspaces) to be able to handle several extensions in the same project.

As the custom extension lives in the repository itself (and we don't want to publish it at npm), what is the best workflow to work?
How should we connect/link the custom extensions with Directus? Both during development and for production.

Note: I am taking into account that now the extensions can be in the root of the extensions folder:
- https://github.com/directus/directus/pull/16822

Thanks!

References:
- https://docs.directus.io/self-hosted/docker-guide.html
- https://github.com/directus/directus/discussions/16989
- https://github.com/directus/directus/discussions/16948
- https://github.com/directus/directus/discussions/14621
- https://github.com/directus/directus/discussions/13968
- https://github.com/directus/directus/discussions/8584

**Comment by abdonrd** [comment]
We just tried with this `Dockerfile`, but it doesn't seem to work:

```dockerfile
FROM node:18-alpine AS builder

WORKDIR /builder

COPY package.json package-lock.json ./
COPY directus-extension-extension-name/ directus-extension-extension-name/
# COPY directus-extension-extension-name-2/ directus-extension-extension-name-2/
RUN npm ci
RUN npm run build

FROM directus/directus:9.22.4

COPY --from=builder /builder/directus-extension-extension-name/ ./extensions/directus-extension-extension-name/
# COPY --from=builder /builder/directus-extension-extension-name-2/ ./extensions/directus-extension-extension-name-2/

CMD npx directus bootstrap && npx directus schema apply snapshots/latest.yaml --yes && npx directus start
```

Note: I am taking into account that now the extensions can be in the root of the extensions folder:
- https://github.com/directus/directus/pull/16822

**Reply by paescuj** [reply]
What exactly isn't working? Is the build failing or the extension not loaded?

**Reply by abdonrd** [reply]
Directus does not load the extension.

**Reply by abdonrd** [reply]
Let me add an screenshot:

<img width="858" alt="Screenshot 2023-01-26 at 12 51 51 PM" src="https://user-images.githubusercontent.com/1007051/214911427-bda49fcb-0475-46be-8b40-45f6509fc59d.png">

**Reply by paescuj** [reply]
Should be more like this (in your posted `Dockerfile` the extension doesn't get build):
```Dockerfile
FROM node:18-alpine AS builder

WORKDIR /builder/directus-extension-extension-name
COPY directus-extension-extension-name/ .
RUN npm ci
RUN npm run build

FROM directus/directus:9.22.4

COPY --from=builder --chown=node:node /builder/directus-extension-extension-name/ ./extensions/directus-extension-extension-name/

CMD npx directus bootstrap && npx directus schema apply snapshots/latest.yml --yes && npx directus start
```

**Reply by abdonrd** [reply]
In this `Dockerfile` https://github.com/directus/directus/discussions/17310#discussioncomment-4789268, the extension is built...

<img width="926" alt="Screenshot 2023-01-30 at 9 42 26 AM" src="https://user-images.githubusercontent.com/1007051/215508201-fa71002e-4099-4a14-b7db-3e95350d27b2.png">

**Reply by paescuj** [reply]
But on the host and not during `docker build`, right? The dist folder should contain the `api.js` & `app.js` files instead...

**Reply by abdonrd** [reply]
I would say not. When I build the extension locally (with `npm ci` & `npm run build`) it just creates an `dist/index.js`:

<img width="930" alt="Screenshot 2023-01-30 at 10 10 05 AM" src="https://user-images.githubusercontent.com/1007051/215515396-e37514b0-7865-4b08-bb6e-1eabd2e04ad6.png">

**Reply by abdonrd** [reply]
If it helps you, I have uploaded it in a branch:
https://github.com/abdonrd/directus-docker-extensions-demo/tree/docker

**Reply by paescuj** [reply]
Sorry, I thought it to be a `bundle` extension, that's why I've expected `api.js` & `app.js` files...

Your problem is actually much simpler:
```diff
diff --git a/docker-compose.yml b/docker-compose.yml
index 3a69128..5019682 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -24,7 +24,6 @@ services:
     ports:
       - 8055:8055
     volumes:
-      - ./extensions:/directus/extensions
       - ./snapshots:/directus/snapshots
       - ./uploads:/directus/uploads
     networks:
```

**Reply by abdonrd** [reply]
Thanks @paescuj, now it loads the extension!

And what would the workflow be like while we are developing the extension? Should we rebuild with each change?
```
docker compose build
docker compose up
```

**Reply by paescuj** [reply]
In the dev environment I'd probably mount the extensions folder into the container, set [`EXTENSIONS_AUTO_RELOAD`](https://docs.directus.io/self-hosted/config-options.html#extensions) to `true` and simultaneously run `npm run build --watch` in the extension.

**Reply by abdonrd** [reply]
Thanks @paescuj!

Now I just created another branch:
https://github.com/abdonrd/directus-docker-extensions-demo/tree/auto-reload

First I move the extension inside the `extensions` folder, and then enable the `EXTENSIONS_AUTO_RELOAD`.

But it doesn't seem to work, the Directus server never re-starts:

https://user-images.githubusercontent.com/1007051/215597158-c90b692b-ff07-4bd2-b40a-fdd7a1ab3bcd.mov

**Reply by paescuj** [reply]
Might be a similar problem as in https://github.com/directus/directus/issues/13675#issuecomment-1144580939 - on some platforms the changes aren't detected inside the docker container.

**Comment by abdonrd** [comment]
This workflow seems to work better for us...

- We have the local extensions with in the `extensions` folder and using npm workspaces:
```json
  "scripts": {
    "build": "npm run build --workspaces --if-present"
  },
  "workspaces": [
    "extensions/directus-extension-*"
  ]
```

- In the `docker-compose.yml` we use the oficial Directus image with a custom command.
```yaml
version: '3'
services:
  database:
    # [...]

  cache:
    # [...]

  directus:
    container_name: directus
    image: directus/directus:9.22.4
    ports:
      - 8055:8055
    volumes:
      - ./extensions:/directus/extensions
      - ./snapshots:/directus/snapshots
      - ./uploads:/directus/uploads
    networks:
      - directus
    depends_on:
      - cache
      - database
    command:
      - /bin/sh
      - -c
      - |
        npx directus bootstrap &&
        npx directus schema apply snapshots/latest.yaml --yes &&
        npx directus start
    environment:
      # [...]

networks:
  directus:
```

- In the `Dockerfile` we build the extensions and copied later:
(and we also use the oficial Directus image with a custom command)
```dockerfile
FROM node:18-alpine AS builder

WORKDIR /builder

COPY package.json package-lock.json ./
COPY extensions/ extensions/
RUN npm ci
RUN npm run build

FROM directus/directus:9.22.4

COPY snapshots/ ./snapshots/
COPY --from=builder /builder/extensions/directus-extension-extension-name/ ./extensions/directus-extension-extension-name/
# COPY --from=builder /builder/extensions/directus-extension-extension-name-2/ ./extensions/directus-extension-extension-name-2/

CMD npx directus bootstrap && npx directus schema apply snapshots/latest.yaml --yes && npx directus start
```

So for the development workflow we run two terminals:
- `npm install`
  `cd extensions/directus-extension-extension-name/` and `npm run dev`

- `docker compose up`

**Reply by abdonrd** [reply]
But the `EXTENSIONS_AUTO_RELOAD` doesn't work. Probably because this: https://github.com/directus/directus/issues/13675#issuecomment-1144580939

**Reply by abdonrd** [reply]
Another update to the `Dockerfile`:

```dockerfile
FROM node:18-alpine AS builder

WORKDIR /builder

COPY package.json package-lock.json ./
COPY extensions/ extensions/
RUN npm ci && \
  npm run build

FROM directus/directus:9.22.4

WORKDIR /directus

COPY snapshots/ ./snapshots/
COPY --from=builder /builder/extensions/directus-extension-extension-name/ ./extensions/directus-extension-extension-name/
# COPY --from=builder /builder/extensions/directus-extension-extension-name-2/ ./extensions/directus-extension-extension-name-2/

CMD npx directus bootstrap && npx directus schema apply snapshots/latest.yaml --yes && npx directus start
```

**Reply by abdonrd** [reply]
Another update to the `Dockerfile`:

```dockerfile
FROM node:18-alpine AS builder

WORKDIR /builder

COPY package.json package-lock.json ./
COPY extensions/ extensions/
RUN npm ci && \
  npm run build

FROM directus/directus:9.22.4

WORKDIR /directus

COPY snapshots/ ./snapshots/
COPY extensions/migrations/ ./extensions/migrations/
COPY --from=builder /builder/extensions/directus-extension-extension-name/ ./extensions/directus-extension-extension-name/
# COPY --from=builder /builder/extensions/directus-extension-extension-name-2/ ./extensions/directus-extension-extension-name-2/

CMD npx directus bootstrap && npx directus schema apply snapshots/latest.yaml --yes && npx directus start
```

**Reply by abdonrd** [reply]
Another update to the `Dockerfile` because the migrations:

```dockerfile
FROM node:18-alpine AS builder

WORKDIR /builder

COPY package.json package-lock.json ./
COPY extensions/ extensions/
RUN npm ci && \
  npm run build

FROM directus/directus:9.23.1

# Switch to user 'node' and directory '/directus'
USER node
WORKDIR /directus

COPY snapshots/ ./snapshots/
COPY extensions/migrations-custom/ ./extensions/migrations-custom/
COPY --from=builder /builder/extensions/directus-extension-extension-name/ ./extensions/directus-extension-extension-name/
# COPY --from=builder /builder/extensions/directus-extension-extension-name-2/ ./extensions/directus-extension-extension-name-2/

CMD npx directus bootstrap && \
  npx directus schema apply snapshots/latest.yaml --yes && \
  mkdir -p extensions/migrations/ && \
  cp -r extensions/migrations-custom/* extensions/migrations/ && \
  npx directus database migrate:latest && \
  npx directus start
```

**Reply by algoflows** [reply]
This is a what everyone who develops with seperate production environment and uses modern CI has been looking for, insane how this is not properly covered in the documentation, spent an entire weekend lost on this. Frustrated with the docs that don't cover this at all. Still many elements are not very clear still.

**Comment by AkioSarkiz** [comment]
Here is my example. I copy the extensions folder and build every folder then just copy that into the directus image. 

```Dockerfile
####################################################################################################
## Build Packages

FROM node:18.20.4-alpine AS builder

WORKDIR /directus

COPY extensions/ extensions/

RUN for ext in $(ls extensions); do \
    echo "Building extension: $ext"; \
    cd extensions/$ext && \
    npm install && \
    npm run build || exit 1; \
    cd -; \
done


####################################################################################################
## Create Production Image
FROM directus/directus:11.1.0 as directus

COPY --from=builder --chown=node:node /directus/extensions /directus/extensions
COPY --chown=node:node ./database /directus/database
COPY --chown=node:node ./snapshots /directus/snapshots
COPY --chown=node:node ./uploads /directus/uploads
```

**Comment by kueronya** [comment]
Just stumbled across this while working on a full docker-compose only configured setup of my service stack and while it's not directly related to OP's question, i thought it may be useful to drop this as another reference.

As of docker compose v2.30.0 there is a [Post-start hook](https://docs.docker.com/compose/how-tos/lifecycle/) which can be used to clone and build extensions right from github on container startup.
It does significantly increase the startup time of the container tho, but i still prefer this over building a new custom image on every directus or extension release.

The relevant parts of the the docker-compose.yml:

```
name: service-stack
services:
  directus:
    image: directus/directus:11.2.2
    post_start:
      - command: >
          /bin/sh -c "
            # Install git
              apk add --no-cache git &&

            # Extension directus-extension-flexible-editor
            # https://github.com/formfcw/directus-extension-flexible-editor
            # Pull latest or clone the repo if it doesn't exist yet
            # This does log an error, as it triggers EXTENSION_AUTO_RELOAD but there is no dist folder yet.
              git -C /directus/extensions/directus-extension-flexible-editor pull || git clone https://github.com/formfcw/directus-extension-flexible-editor.git /directus/extensions/directus-extension-flexible-editor &&
            # Install dependencies with production=false to include devDependencies (devDependencies needed for building the extension)
              npm install --prefix /directus/extensions/directus-extension-flexible-editor --production=false &&
            # Build the extension
              npm run build --prefix /directus/extensions/directus-extension-flexible-editor &&
            # Touch the package.json to trigger EXTENSION_AUTO_RELOAD once again (this time with dist folder in place)
              touch /directus/extensions/directus-extension-flexible-editor/package.json

            # Repeat for other extensions
            # Extension directus-extension-generate-types
            # https://github.com/maltejur/directus-extension-generate-types
              git -C /directus/extensions/directus-extension-generate-types pull || git clone https://github.com/maltejur/directus-extension-generate-types.git /directus/extensions/directus-extension-generate-types &&
              npm install --prefix /directus/extensions/directus-extension-generate-types --production=false &&
              npm run build --prefix /directus/extensions/directus-extension-generate-types &&
              touch /directus/extensions/directus-extension-generate-types/package.json
          "
        # Run all of this as root
        user: root
    volumes:
      - ./data/directus/extensions:/directus/extensions
    environment:
      EXTENSIONS_AUTO_RELOAD: "true"

```

Ref https://github.com/directus/directus/discussions/16989 (unfortunately locked)
